### PR TITLE
Master project rating improvements dbh

### PR DESCRIPTION
**Pull request by dbh-odoo**
Task: https://www.odoo.com/web#id=30903&view_type=form&model=project.task&action=333&active_id=131&menu_id=4720

Pad: https://pad.odoo.com/p/r.9dde16410e8daf3ad9792a0b790d373e

Description of the issue/feature this PR addresses:

Current behavior before PR:

Desired behavior after PR is merged:

--
I confirm I have signed the CLA and read the PR guidelines at www.odoo.com/submit-pr

### CODE REVIEW

File: addons/rating/models/rating.py [hunk 1]
```diff
@@ -78,6 +98,16 @@ def action_open_rated_object(self):
             'views': [[False, 'form']]
         }
 
+    def get_parent_model_name(self, vals):
```
Comment (jem-odoo) [inline addons/rating/models/rating.py]: Those methods should be on the rating.mixin

File: addons/rating/models/rating.py [hunk 2]
```diff
@@ -78,6 +98,16 @@ def action_open_rated_object(self):
...
+        """ Return the parent resource model name. The dict of values passed to
+            ``create`` when a record of this model is created."""
+        return None
+
+    def get_parent_res_id_value(self):
+        """ Return value of Parent Resource Id, or to write on the Parent Resource Id after its
+            creation."""
+        return {'parent_res_id': False}
```
Comment (jem-odoo) [inline addons/rating/models/rating.py]: Called it `rating_get_parent_id`, and return only the id

File: addons/rating/models/rating.py (hunk 1 above)
Comment (jem-odoo) [inline addons/rating/models/rating.py]: Called it `rating_get_parent_model_name`

File: addons/rating/models/rating.py [hunk 3]
```diff
@@ -32,16 +32,27 @@ def new_access_token(self):
...
+    parent_res_name = fields.Char(compute='_compute_parent_res_name', String='Parent Document Name', store=True)
+    parent_res_model_id = fields.Many2one('ir.model', 'Parent Related Document Model', index=True)
+    parent_res_model = fields.Char(string='Parent Document Model', store=True, related='parent_res_model_id.model', index=True)
+    parent_res_id = fields.Integer(string='Parent Document ID', index=True)
     rated_partner_id = fields.Many2one('res.partner', string="Rated person", help="Owner of the rated resource")
     partner_id = fields.Many2one('res.partner', string='Customer', help="Author of the rating")
     rating = fields.Float(string="Rating", group_operator="avg", default=0, help="Rating value: 0=Unhappy, 10=Happy")
     rating_image = fields.Binary('Image', compute='_compute_rating_image')
-    rating_text = fields.Char(string='Rating', compute='_compute_rating_text')
+    rating_text = fields.Char(string='Rating', store=True, compute='_compute_rating_text')
```
Comment (jem-odoo) [inline addons/rating/models/rating.py]: You cannot store translated terms.
In a separated commit, try to group on rating_text, but change this field into a selection computed field (readonly)

File: addons/rating_project/views/project_view.xml [hunk 4]
```diff
@@ -23,6 +23,9 @@
                     <separator />
                     <filter string="Rating for tasks" name="rating_tasks" domain="[('res_model', '=', 'project.task')]" />
                 </xpath>
+                <xpath expr="//filter[@name='Responsible']" position="before">
```
Comment (jem-odoo) [inline addons/rating_project/views/project_view.xml]: No capital letter in `name` attribute

Comment (jem-odoo) [inline addons/rating_project/views/project_view.xml]: Also make all inherited views in rating_project `primary`. Since rating is not only used in project, we don't want to see 'Rating for task' when we are looking at livechat rating.
Adapt ir.actions accordingly.

File: addons/rating/views/rating_view.xml [hunk 5]
```diff
@@ -24,7 +24,9 @@
                     <sheet>
                         <group>
                             <group>
-                                <field name="res_name"/>
+                                <field name="res_model" invisible="1" />
+                                <field name="res_id" invisible="1" />
+                                <field name="res_name" widget="link_record" options="{'res_model_field': 'res_model', 'res_id_field': 'res_id'}" />
```
Comment (jem-odoo) [inline addons/rating/views/rating_view.xml]: I bet an action (in python) that return the good res_model/res_id can make it. Is a new widget really required ?

File: addons/rating/models/rating.py [hunk 6]
```diff
@@ -57,7 +68,16 @@ def _compute_rating_image(self):
...
-            rating.rating_text = text[rating.rating] or _('No rating yet')
+            rating.rating_text = rating.rating and text[rating.rating] or _('No rating yet')
+
+    @api.model
+    def create(self, vals):
+        record = super(Rating, self).create(vals)
+        res = self.env[record.res_model].browse(record.res_id)
+        record.sudo().write(res.get_parent_res_id_value())
+        parent_res_model_name = res.get_parent_model_name(vals)
+        record.parent_res_model_id = self.env['ir.model'].search([('model', '=', parent_res_model_name)]).id
```
Comment (jem-odoo) [inline addons/rating/models/rating.py]: Rather than make a `write` rigth after the `create`, shouldn't it be better to update the values given to the create directly ? Like it is done in mail_alias.py

File: addons/rating/models/rating.py [hunk 7]
```diff
@@ -55,9 +70,20 @@ def _compute_rating_image(self):
     @api.multi
     @api.depends('rating')
     def _compute_rating_text(self):
-        text = {10: _('Satisfied'), 5: _('Not satisfied'), 1: _('Highly dissatisfied')}
+        text = {10: _('satisfied'), 5: _('not_satisfied'), 1: _('highly_dissatisfied')}
```
Comment (jem-odoo) [inline addons/rating/models/rating.py]: Don't translate the terms, as they are technical !

Comment (jem-odoo) [inline addons/rating/models/rating.py]: I also think we should use the same repartition than in `rating_get_grades` (see https://github.com/odoo/odoo/blob/master/addons/rating/models/rating.py#L237)

File: addons/rating/models/rating.py [hunk 8]
```diff
@@ -55,9 +70,20 @@ def _compute_rating_image(self):
...
         for rating in self:
-            rating.rating_text = text[rating.rating] or _('No rating yet')
+            rating.rating_text = rating.rating and text[rating.rating] or _('no_rating')
+
+    @api.model
+    def create(self, vals):
+        if vals.get('res_model_id') and vals.get('res_id'):
+            model_name = self.env['ir.model'].browse(vals['res_model_id']).model
```
Comment (jem-odoo) [inline addons/rating/models/rating.py]: Is it possible that this could trigger 'Access Error' ? Check if every user can read `ir.model`. Otherwise, a `sudo()` will be required.

File: addons/rating_project/views/project_view.xml [hunk 9]
```diff
@@ -32,7 +36,7 @@
             <field name="inherit_id" ref="project.view_task_form2"/>
             <field name="arch" type="xml">
                 <xpath expr="//button[@name='toggle_active']" position="before">
-                    <button name="%(rating_rating_action_task)d" type="action"
+                    <button name="action_open_rating_view" type="object"
```
Comment (jem-odoo) [inline addons/rating_project/views/project_view.xml]: Why this change ? Can't you keep action defined in xml (and eventually modify it) ?

Comment (atp-odoo) [inline addons/rating_project/views/project_view.xml]: no.. we can't use same exisitng action defined in xml. because we already set primary mode view based on product view action..(which has same model but res_id and res_model_id area different.)
